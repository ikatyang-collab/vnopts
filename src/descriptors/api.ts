import { Descriptor } from '../types.js'

export const apiDescriptor: Descriptor = {
  key: key =>
    /^[$_a-zA-Z][$_a-zA-Z0-9]*$/.test(key) ? key : JSON.stringify(key),
  value(value) {
    if (value === null || typeof value !== 'object') {
      return JSON.stringify(value)
    }

    if (Array.isArray(value)) {
      return `[${value
        .map(subValue => apiDescriptor.value(subValue))
        .join(', ')}]`
    }

    const keys = Object.keys(value)
    return keys.length === 0
      ? '{}'
      : `{ ${keys
          .map(
            key =>
              `${apiDescriptor.key(key)}: ${apiDescriptor.value(value[key])}`,
          )
          .join(', ')} }`
  },
  pair: ({ key, value }) => apiDescriptor.value({ [key]: value }),
}
