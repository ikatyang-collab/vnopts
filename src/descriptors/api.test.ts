import { expect, test } from 'vitest'
import { apiDescriptor } from './api.js'

test('apiDescriptor', () => {
  expect(apiDescriptor.value({})).toMatchSnapshot()
  expect(apiDescriptor.value('key')).toMatchSnapshot()
  expect(apiDescriptor.value({ key: 'value' })).toMatchSnapshot()
})
