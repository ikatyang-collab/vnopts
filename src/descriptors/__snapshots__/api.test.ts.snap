// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`apiDescriptor 1`] = `"{}"`;

exports[`apiDescriptor 2`] = `""key""`;

exports[`apiDescriptor 3`] = `"{ key: "value" }"`;
